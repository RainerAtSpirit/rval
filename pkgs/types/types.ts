import { _isPlainObject } from "@r-val/core";

export type TypeChecker<T> = (value: any) => T

function createTypeChecker<T>(predicate: (v: any) => boolean, typeName: string): TypeChecker<T> {
  return Object.assign(
    function(v) {
      if (!predicate(v)) {
        const vAsStr = _isPlainObject(v) || Array.isArray(v) ? JSON.stringify(v) : v
        throw new Error(`Typecheck failed, expected '${typeName}', got: (${typeof v}) '${vAsStr}'`)
      }
      return v
    },
    {
      typeName: typeName,
    }
  )
}

function getTypeName(fn) {
  if (fn.typeName) return fn.typeName
  throw new Error('Not a type-checker function:' + fn)
}

function isA(typeChecker, value) {
  try {
    typeChecker(value)
    return true
  } catch (e) {
    return false
  }
}

export const isLiteral = <T>(value: T) => createTypeChecker<T>(v => v === value, "" + value)
export const isNull = isLiteral(null)
export const isUndefined = isLiteral(undefined)
export const isNumber = createTypeChecker<number>(v => typeof v === 'number', 'number')
export const isString = createTypeChecker<number>(v => typeof v === 'string', 'string')
export const isBoolean = createTypeChecker<number>(v => typeof v === 'boolean', 'boolean')
export const isFunction = createTypeChecker<Function>(v => typeof v === 'function', 'function')
export const isDate = isInstanceOf(Date)

export function isInstanceOf<T>(type: new (...args: any[]) => T): TypeChecker<T> {
  return createTypeChecker<T>(v => v instanceof type, type.name)
}

// Could use typings, but requires a lot of overloads
export function isUnion<T>(...types: TypeChecker<any>[]): TypeChecker<T> {
  return createTypeChecker(
    v => types.some(t => isA(t, v)),
    types.map(getTypeName).join(" | ")
  )
}

export function isMaybe<T>(subTypeChecker: TypeChecker<T>): TypeChecker<T|null|undefined> {
  return isUnion(subTypeChecker, isUndefined, isNull)
}

// Could use typings, but requires a lot of overloads
export function isEnum<T>(values: T[]): TypeChecker<T> {
  return isUnion(...values.map(isLiteral))
}

export function isArray<T>(subTypeChecker: TypeChecker<T>): TypeChecker<T[]> {
  return createTypeChecker(
    v => Array.isArray(v) && v.every(i => isA(subTypeChecker, i)),
    `Array<${getTypeName(subTypeChecker)}>`
  )
}

export function isMap<T>(subTypeChecker: TypeChecker<T>): TypeChecker<{[key: string]: T}> {
  return createTypeChecker(
    v => _isPlainObject(v) && Object.keys(v).every(key => isA(subTypeChecker, v[key])),
    `Map<${getTypeName(subTypeChecker)}>`
  )
}

export function isShape<T extends {[key: string]: TypeChecker<any> }>(shape: T): TypeChecker<T> {
  const typeString = "Object<{\n\t"
    + Object.keys(shape).map(key => `${key}: ${getTypeName(shape[key])}`).join(",\n\t")
    + "\n}>";
  return createTypeChecker(
    v => Object.keys(shape).every(key => isA(shape[key], v[key])),
    typeString
  )
}

export function invariant<T>(predicate: (v: T) => boolean): TypeChecker<T> {
  return createTypeChecker(predicate, `invariant '${predicate.toString()}'`)
}
