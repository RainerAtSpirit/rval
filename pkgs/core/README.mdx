---
name: Core api
order: 1
menu: API
route: /api/core
---

# @r-val/core

`@r-val/core` package contains all the core functionality of RVal. 
The most important concepts are explain on the [getting started](#/introduction/the-basics) page.

## `val`

**Signature** 

`val(initialValue, preProcessor?) -> reactive value`

**Description**

`val` creates a reactive value that can be used to store state.
The `initialValue` froms the initial state of the reative value.

`val` returns a function that takes zero or one arguments:
* If the function is invoked without arguments, the current value is returned
* If the function is invoked with one argument, and that value is not a function, the argument becomes the new state of the reactive value.
* If the function is invoked with a function as argument, that function is invoked and should produce the new state, based on the current state. Built-in update functions can be found in the [`@r-val/updaters` package](#/api/updaters).

```javascript
const temperature = val(37)
console.log(temperature()) // prints 37

temperature(38)
console.log(temperature()) // prints 38

temperature(v => v + 1)
console.log(temperature()) // prints 39
```

For the `preProcessor` argument, see [#preprocessors].

## PreProcessors

By default, any new value state stored in a `val` is automatically deeply frozen, even the value is a plain array or object.

The `preProcessor` argument to apply some preprocessing on the values that are being stored, or perform validations.
The signature of a pre-processor is: `(newValue, currentValue, rvalInstance) -> newValue`. Where:
 * `newValue` the value that is about to be stored
 * `currentValue` the value that is currently stored
 * `rvalInstance` the current RVal instance (see the `rval()` function)
 * The return value is the value that will be passed to the next pre-processor, or if there is none, that will be stored.

It is possible to chain multiple pre processors, by simply passing an array of preprocessors to `val`.

`val` uses pointer equality to detect and propagate updates. So if your previous detects that the current update can or needs to be aborted, just `return currentValue`.

For example, the following reactive value `profit` has two pre-processors. 
The first one automatically converts new values to strings,
the second one performs some validation, refusing to store numbers that 
are smaller then the previous value.

```javascript
function convertToNumber(newValue) {
    if (typeof newValue === "number")
        return newValue
    else
        return parseFloat(newValue)
}

const profit = val(0, [
    // convert input to number
    convertToNumber,
    // check if profits increase
    (newValue, currentValue) => {
        if (newValue < currentValue)
            throw new Error("Invariant failed! Profits should increase")
        return newValue
    }
])

profit(-5) // Throws exception: Profits should increase

profit("7") // Parses the number, and sets profit to number 7
```



## `drv`

**Signature**

`drv(derivation, setter?) -> reactive value`

**Description**

## `sub`

**Signature**

* `sub(source, listener, options) -> disposer` &mdash; the default way to create a subscription to an reactive `source`
* `sub(listener, options) -> (source) -> disposer` &mdash; create a curried subscription, 

**Subscription `options`:**

* `fireImmediately` (default `false`). `true`: the listener should be immediately invoked, or `false`: only when the first _change_ is observed in `source`.

**Listener arguments:**

* `value`
* `previousValue`

**Description**

TODO

The curried version of the api can be useful to create utilities, for example a `logger` that prints the values of any state transition:

```javascript
const log = sub((current, previous) => {
    console.log(`${previous} -> ${current}`)
})

log(someReactiveValue)
log(someOtherReactiveValue)
```

## `act`

## `effect`

## `isVal`

## `isDrv`

## `configure`

**Signature**

`configure(options)`

**Configuration options**

* `deepFreeze`: automatically deeply freezes any values arrays or plain objects that are stored in a reactive value. Best enabled in development enviroments. Default: `true`. Should be disabled in production enviroments for optimal performance.

Set's the the options for the current RVal instance (see also `rval` and `defaultInstance`).


## `rval`

Signature: 

* `rval()` &mdash; creates a _new_ RVal instance
* `rval(reactiveValue)` &mdash; returns the RVal instances that created the given reactive value.

Returns an rval instance, that exposes the APIs:

* `val`
* `drv`
* `sub`
* `act`
* `effect`
* `configure`

## `defaultInstance`

The default RVal instance used by all the top-level api's exposed by the module. 
For more info about what an RVal instance is, see `rval`.

## `$RVal`
