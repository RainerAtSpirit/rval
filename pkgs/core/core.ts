export type Listener<T = any> = (value: T) => void

export type Thunk = () => void

export type Disposer = Thunk

// Exposed, this might be useful for third part reflection based (dev) tools
export const $RVal = typeof Symbol === "undefined" ? "$RVal" : Symbol.for('$RVal')

export interface Observable<T = unknown> {
  (): T
}

export interface Drv<T = unknown> extends Observable<T> {}

export interface Val<T = unknown, S = T> extends Observable<T> {
  (newValue: T | S): void
}

interface RValContext {
  isUpdating: boolean
  pending: Thunk[],
  currentlyComputingStack: Set<ObservableAdministration>[]
  currentlyComputing: Set<ObservableAdministration>
  isRunningReactions: boolean
  runPendingObservers()
}

interface ObservableAdministration {
  addListener(observer: Thunk)
  removeListener(observer: Thunk)
  get(): any
}

// TODO: swap types of S, T, infer
// also for Val
export type PreProcessor<T = unknown, S = T> = (newValue: T | S, baseValue?: T, api?: RValFactories) => T

export interface RValFactories {
  val<T, S>(initial: S, preProcessor: PreProcessor<T, S>): Val<T, S>
  val<T>(initial: T): Val<T, T>
  drv<T>(derivation: () => T): Drv<T>
  sub<T>(
    src: Observable<T>,
    listener: Listener<T>,
    options?: SubscribeOptions
  ): Disposer
  effect<T>(fn: () => T, onInvalidate: (onChanged: () => boolean, pull: () => T) => void): Thunk
  batch<R>(updater: () => R): R
  batched<T extends Function>(fn: T): T
}

export interface SubscribeOptions {
  fireImmediately?: boolean
  scheduler?: (run: Thunk) => void
}

const NOT_TRACKING = 0
const STALE = 1
const UP_TO_DATE = 2

export function rval(base?: Val<any, any>): RValFactories {
  if (arguments.length) {
    if (!isVal(base) && !isDrv(base))
      throw new Error("Expected val as first argument to rval")
    return (base[$RVal] as any).api
  }
  const context: RValContext = {
    isUpdating : false,
    pending: [],
    currentlyComputingStack: [],
    get currentlyComputing() {
      return this.currentlyComputingStack[this.currentlyComputingStack.length - 1]
    },
    isRunningReactions: false,
    runPendingObservers
  }

  function runAfterBatch(t: Thunk) {
    context.pending.push(t)
    if (!context.isUpdating) runPendingObservers(); // TODO: is this line ever hit?
  }

  function val<T, S>(initial: S, preProcessor = defaultPreProcessor): Val<T, S> {
    return new ObservableValue(context, api, initial, preProcessor).get as any
  }

  function drv<T>(derivation: () => T): Drv<T> {
    return new Computed<T>(context, api, derivation).get as any
  }

  function effect<T>(fn: () => T, onInvalidate: (onChanged: () => boolean, pull: () => T) => void): Thunk {
    // TODO: avoid double wrapping computeds
    const computed = new Computed(context, api, fn)
    let scheduled = true
    let disposed = false
    
    function didChange() {
      if (disposed) return false
      const changed = computed.someDependencyHasChanged()
      if (!changed) {
        scheduled = false // no pull is expected
      }
      return changed
    }
    function pull() {
      if (disposed) {
        throw new Error("[rval] pulling from already disposed effect")
      }
      scheduled = false
      return computed.get()
    }
    function onDirty () {
      if (scheduled || disposed) return
      scheduled = true
      runAfterBatch(() => onInvalidate(didChange, pull))
    }
    
    computed.addListener(onDirty)
    onInvalidate(didChange, pull)
    return once(() => {
      disposed = true
      computed.removeListener(onDirty)
    })
  }

  function sub<T>(
    src: Observable<T>,
    listener: Listener<T>,
    options?: SubscribeOptions
  ): Disposer {
    let lastSeen: T | undefined = undefined
    let firstRun = true
    const effectDisposer = effect(src, (didChange, pull) => {
      if (didChange()) {
        const v = pull()
        if (!firstRun && v !== lastSeen) listener(v)
        lastSeen = v
        firstRun = false
      }
    })
    return () => {
      effectDisposer()
    }
  }

  // TODO: autowrap with update and warn?
  function batch<R>(updater: () => R): R {
    let prevUpdating = context.isUpdating
    context.isUpdating = true
    try {
      return updater()
    } finally {
      context.isUpdating = prevUpdating
      if (!context.isUpdating) {
        runPendingObservers()
      }
    }
  }

  // TODO: rename to act, and kill batch
  function batched<T extends Function>(fn: T): T {
    return (function updater(this: any) {
      const self = this
      // TODO: optimize: short-circuit if already in transaction
      return batch(() => fn.apply(self, arguments))
    } as any) as T
  }

  function runPendingObservers() {
    if (!context.isUpdating && !context.isRunningReactions) {
      context.isRunningReactions = true
      while (context.pending.length) {
        // N.B. errors here cause other pending subscriptions to be aborted!
        context.pending.splice(0).forEach(run)
      }
      context.isRunningReactions = false
    }
  }

  // prettier-ignore
  const api = { val, drv, sub, batch, batched, effect }
  return api
}

const defaultPreProcessor = value => value
export const defaultContext = rval()

class ObservableValue<T> implements ObservableAdministration {
  listeners: Thunk[] = []
  value: T
  constructor(private context: RValContext, public api: RValFactories, state: T, private preProcessor) {
    this.get = this.get.bind(this)
    hiddenProp(this.get, $RVal, this)
    this.value = deepfreeze(preProcessor(state, undefined, this.api)) // TODO: make freeze an option
  }
  addListener(listener) {
    this.listeners.push(listener)
  }
  removeListener(listener) {
    removeCallback(this.listeners, listener)
  }
  get(newValue?: T) {
    switch (arguments.length) {
      case 0:
        registerRead(this.context, this)
        return this.value
      case 1:
      // prettier-ignore
        if (this.context.currentlyComputing) throw new Error('derivations cannot have side effects and update values')
        // if (!isUpdating)
        //   throw new Error("val can only be updated within an 'update' context") // TODO: make ok, but optionally support / enforce batching
        newValue = this.preProcessor(newValue, this.value, this.api)
        if (newValue !== this.value) {
          this.value = deepfreeze(newValue) // TODO: make freeze an option
          batch(() => { // optimize: no need to wrap if already in transaction
            runAll(this.listeners)
          })
        }
        // TODO: return this.value ?
        break
      default:
        throw new Error('val expects 0 or 1 arguments')
    }
  }
}

class Computed<T = any> implements ObservableAdministration {
  listeners: Thunk[] = []
  inputValues: any[] | undefined = undefined
  observing!: Set<ObservableAdministration>
  state = NOT_TRACKING
  dirtyCount = 0
  value: T = undefined!
  constructor(private context: RValContext, public api: RValFactories, public derivation: () => T) {
    this.get = this.get.bind(this)
    hiddenProp(this.get, $RVal, this)
  }
  markDirty = () => {
    if (++this.dirtyCount === 1) {
      this.state = STALE
      runAll(this.listeners)
    }
  }
  addListener(observer) {
    this.listeners.push(observer)
  }
  removeListener(observer) {
    removeCallback(this.listeners, observer)
    if (!this.listeners.length) {
      this.observing.forEach(o => o.removeListener(this.markDirty))
      this.value = undefined!
      this.state = NOT_TRACKING
      this.inputValues = undefined
    }
  }
  registerDependency(sub: ObservableAdministration) {
    this.observing.add(sub)
  }
  someDependencyHasChanged() {
    switch(this.state) {
      case NOT_TRACKING: return true
      case UP_TO_DATE: return false
      case STALE: 
        // TODO: did should be done in tracking context, otherwise
        // deps are registered double?
        if (!inputSetHasChanged(this.observing, this.inputValues)) {
          this.dirtyCount = 0
          this.state = UP_TO_DATE
          return false;
        }
      }
      return true;
  }
  track() {
    if (!this.someDependencyHasChanged()) return
    const oldObserving = this.observing
    const [newValue, newObserving] = track(this.context, this.derivation)
    this.value = newValue
    this.observing = newObserving
    registerDependencies(this.markDirty, oldObserving, newObserving)
    this.inputValues = recordInputSet(newObserving)
    this.dirtyCount = 0
    this.state = UP_TO_DATE
  }
  get() {
    // console.log("GET - "+ this.derivation.toString())
    // something being computed? setup tracking
    registerRead(this.context, this)
    // yay, we are up to date!
    if (this.state === UP_TO_DATE) return this.value
    // nope, we are not, and no one is observing either
    if (!this.context.currentlyComputing && !this.listeners.length)
      return this.derivation()
    // maybe scheduled, definitely tracking, value is needed, track now!
    this.track()
    return this.value
  }
}

type DependencySet = (ObservableAdministration | any)[]

function track<R>(context: RValContext, fn: () => R): [R, Set<ObservableAdministration>] {
  const observing = new Set()
  context.currentlyComputingStack.push(observing)
  const res = fn()
  context.currentlyComputingStack.pop()
  return [res, observing]
}

function registerDependencies(listener: Thunk, oldDeps: Set<ObservableAdministration>, newDeps: Set<ObservableAdministration>) {
  // Optimize: 
  if (!oldDeps) {
    newDeps.forEach(d => d.addListener(listener))
  } else {
    newDeps.forEach(o => {
      if (!oldDeps.has(o)) o.addListener(listener)
    })
    oldDeps.forEach(o => {
      if (!newDeps.has(o)) o.removeListener(listener)
    })
  }
}

function registerRead(context: RValContext, observable: ObservableAdministration) {
  // optimize: same last touched by optimization as MobX
  // Sets are used, and keep insertion order, which is important for optimal performance! 
  // (to make someDependencyHasChanged cheap and not
  // re-evaluate deps that might not be needed in the future due some branching logic) 
  if (context.currentlyComputing) context.currentlyComputing.add(observable)
}

function recordInputSet(deps: Set<ObservableAdministration>): any[] {
  // optimize: write more efficiently
  return Array.from(deps).map(currentValue)
}

function inputSetHasChanged(deps: Set<ObservableAdministration>, inputs?: any[]) {
  return !deps || !inputs || !Array.from(deps.values()).every((o, idx) => o.get() === inputs[idx])
}

function currentValue(dep: ObservableAdministration): any {
  // Returns the current, last known (computed) value of a dep
  // Regardless whether that is stale or not 
  return (dep as any).value
}

function runAll(fns: Thunk[]): void {
  fns.forEach(run)
}

function run(fn: Thunk): void {
  fn()
}

function removeCallback(fns: Thunk[], fn: Thunk) {
  fns.splice(fns.indexOf(fn), 1) // TODO: defensive index check?
}

export function toJS(value) {
  // convert, recursively, all own enumerable, primitive + vals values
}

export function isVal(value: any): value is Val {
  return typeof value === "function" && value[$RVal] instanceof ObservableValue
}

export function isDrv(value: any): value is Drv {
  return typeof value === "function" && value[$RVal] instanceof Computed
}

function once<T extends Function>(fn: T): T {
  // based on 'once' package, but made smaller
  var f: any = function(this: any) {
    if (f.called) return f.value
    f.called = true
    return (f.value = fn.apply(this, arguments))
  }
  f.called = false
  return f
}

export function deepfreeze(o) {
  // based on 'deepfreeze' package, but copied here to simplify build setup :-/
  if (o === Object(o)) {
    Object.isFrozen(o) || Object.freeze(o)
    Object.getOwnPropertyNames(o).forEach(function(prop) {
      prop === 'constructor' || deepfreeze(o[prop])
    })
  }
  return o
}

function hiddenProp(target, key, value) {
  Object.defineProperty(target, key, {
    // N.B.: quoting is important, to prevent minification issue. See keep_quoted option!
    "configurable": true,
    "enumerable": false,
    "writable": false,
    "value": value
  })
}

export const val = defaultContext.val
export const drv = defaultContext.drv
export const sub = defaultContext.sub
export const batch = defaultContext.batch
export const batched = defaultContext.batched
export const effect = defaultContext.effect

