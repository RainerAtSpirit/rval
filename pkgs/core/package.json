{
  "name": "@r-val/core",
  "private": false,
  "version": "0.0.1",
  "description": "",
  "main": "dist/core.js",
  "umd:main": "dist/core.umd.js",
  "unpkg": "dist/core.umd.js",
  "module": "dist/core.mjs",
  "jsnext:main": "dist/core.mjs",
  "react-native": "dist/core.mjs",
  "browser": {
    "./dist/core.js": "./dist/core.js",
    "./dist/core.mjs": "./dist/core.mjs"
  },
  "source": "core.ts",
  "scripts": {},
  "author": "Michel Weststrate",
  "license": "MIT",
  "dependencies": {},
  "files": ["*.ts", "dist"],
  "devDependencies": {},
  "mangle": {
    "reserved": [
      "val",
      "drv",
      "sub",
      "batch",
      "batched",
      "deepfreeze",
      "isVal",
      "isDrv",
      "defaultContext",
      "configure",
      "autoFreeze"
    ]
  }
}
