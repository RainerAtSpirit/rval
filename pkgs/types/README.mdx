---
name: Type preprocessors
order: 3
menu: API
route: /api/types
---

# @r-val/types
