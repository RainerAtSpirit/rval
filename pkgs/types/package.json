{
  "name": "@r-val/types",
  "private": false,
  "version": "0.0.4",
  "description": "",
  "main": "dist/types.js",
  "umd:main": "dist/types.umd.js",
  "unpkg": "dist/types.umd.js",
  "module": "dist/types.mjs",
  "jsnext:main": "dist/types.mjs",
  "react-native": "dist/types.mjs",
  "browser": {
    "./dist/types.js": "./dist/types.js",
    "./dist/types.mjs": "./dist/types.mjs"
  },
  "source": "types.ts",
  "scripts": {},
  "author": "Michel Weststrate",
  "license": "MIT",
  "files": [
    "*.ts",
    "dist"
  ],
  "reserved": [
    "model",
    "mapOf",
    "arrayOf",
    "invariant",
    "toJS"
  ],
  "mangle": {
    "reserved": [
      "configurable",
      "enumerable",
      "writable",
      "value",
      "$RVal",
      "rval",
      "defaultInstance",
      "isVal",
      "isDrv",
      "_deepfreeze",
      "_once",
      "val",
      "drv",
      "sub",
      "run",
      "act",
      "effect",
      "configure",
      "autoFreeze",
      "fireImmediately",
      "model",
      "mapOf",
      "arrayOf",
      "invariant"
    ]
  },
  "gitHead": "a824fe9bd83228fb126c8c6188ef7b85fc2ffdd7"
}