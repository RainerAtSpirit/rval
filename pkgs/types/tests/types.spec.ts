import { val } from "@r-val/core"
import * as t from "@r-val/types"

test('isLiteral', () => {
  const v = val<any>(3, t.isLiteral(3))

  expect(() => v(4)).toThrow("Typecheck failed, expected '3', got: (number) '4'")
  expect(v()).toBe(3)
  v(3)
  expect(v()).toBe(3)
})


test('isNull', () => {
  const v = val<any>(null, t.isNull)

  expect(() => v(undefined)).toThrow("Typecheck failed, expected 'null', got: (undefined) 'undefined'")
  expect(v()).toBe(null)
  v(null)
  expect(v()).toBe(null)
})

test('isUndefined', () => {
  const v = val<any>(undefined, t.isUndefined)

  expect(() => v(null)).toThrow("Typecheck failed, expected 'undefined', got: (object) 'null'")
  expect(v()).toBe(undefined)
  v(undefined)
  expect(v()).toBe(undefined)
})

test('isNumber', () => {
  const v = val<any>(3, t.isNumber)

  expect(() => v("4")).toThrow("Typecheck failed, expected 'number', got: (string) '4'")
  expect(v()).toBe(3)
  v(4)
  expect(v()).toBe(4)
})

test('isString', () => {
  const v = val<any>("3", t.isString)

  expect(() => v(4)).toThrow("Typecheck failed, expected 'string', got: (number) '4'")
  expect(v()).toBe("3")
  v("4")
  expect(v()).toBe("4")
})

test('isBoolean', () => {
  const v = val<any>(false, t.isBoolean)

  expect(() => v(4)).toThrow("Typecheck failed, expected 'boolean', got: (number) '4'")
  expect(v()).toBe(false)
  v(true)
  expect(v()).toBe(true)
})

test('isFunction', () => {
  const v = val<any>(() => {}, t.isFunction)

  expect(() => v(4)).toThrow("Typecheck failed, expected 'function', got: (number) '4'")
  expect(v()).toBeInstanceOf(Function)
  v(() => () => {})
})

test('isInstanceOf', () => {
  class X {

  }

  const v = val<any>(new X, t.isInstanceOf(X))

  expect(() => v({})).toThrow("Typecheck failed, expected 'X', got: (object) '{}'")
  expect(v()).toBeInstanceOf(X)
  v(new X)
})

test('isDate', () => {
  const v = val<any>(new Date, t.isDate)

  expect(() => v(4)).toThrow("Typecheck failed, expected 'Date', got: (number) '4'")
  expect(v()).toBeInstanceOf(Date)
  v(new Date)
})  

test('isMaybeDate', () => {
  const v = val<any>(new Date, t.isMaybe(t.isDate))

  expect(() => v(4)).toThrow("Typecheck failed, expected 'Date | undefined | null', got: (number) '4'")
  expect(v()).toBeInstanceOf(Date)
  v(new Date)
  expect(v()).toBeInstanceOf(Date)
  v(undefined)
  expect(v()).toBe(undefined)
  v(null)
  expect(v()).toBe(null)
})  

test('isEnum', () => {
  const v = val<any>("stop", t.isEnum(["stop", "start"]))

  expect(() => v(4)).toThrow("Typecheck failed, expected 'stop | start', got: (number) '4'")
  expect(v()).toBe("stop")
  v("start")
  expect(v()).toBe("start")
})

test('isArray', () => {
  const v = val<any>(["stop"], t.isArray(t.isEnum(["stop", "start"])))

  expect(() => v(["start", 4])).toThrow(`Typecheck failed, expected 'Array<stop | start>', got: (object) '["start",4]'`)
  expect(v()).toEqual(["stop"])
  v(["stop", "start"])
  expect(v()).toEqual(["stop", "start"])
  v([])
})

test('isMap', () => {
  const v = val<any>({x:"stop"}, t.isMap(t.isEnum(["stop", "start"])))

  expect(() => v({y: "stop", x: 4})).toThrow(`Typecheck failed, expected 'Map<stop | start>', got: (object) '{"y":"stop","x":4}'`)
  expect(v()).toEqual({x:"stop"})
  v({y: "stop", x: "stop"})
  expect(v()).toEqual({y: "stop", x: "stop"})
  v({})
})

describe('isShape', () => {
  const todoShape = t.isShape({
    done: t.isBoolean,
    title: t.isString
  })

  const todoStoreShape = t.isShape({
    todos: t.isArray(todoShape)
  })

  test('simple', () => {
    const t1 = val({ done: false, title: "get coffee"}, todoShape)
    expect(() => t1({} as any)).toThrowErrorMatchingSnapshot()
    expect(() => t1({ done: true, title: 3 } as any)).toThrowErrorMatchingSnapshot()
    t1({
      extra: 7,
      done: false,
      title: "stuff"
    } as any)
  })

  test('complex', () => {
    const store = val({
      todos: [{ done: false, title: "get coffee"}]
    }, todoStoreShape)

    expect(() => store({} as any)).toThrowErrorMatchingSnapshot()
    store({
      todos: []
    })
    expect(() => store({
      todos: [{ done: false, title: "get coffee"}, 3]
    } as any)).toThrowErrorMatchingSnapshot()
  })
})

test('invariant', () => {
  const x = val(3, t.invariant(p => p > 0))

  x(2)
  expect(() => x(-1)).toThrow("Typecheck failed, expected 'invariant 'p => p > 0'', got: (number) '-1'")
})
