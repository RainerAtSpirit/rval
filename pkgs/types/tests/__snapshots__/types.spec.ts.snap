// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isShape complex 1`] = `
"Typecheck failed, expected 'Object<{
	todos: Array<Object<{
	done: boolean,
	title: string
}>>
}>', got: (object) '{}'"
`;

exports[`isShape complex 2`] = `
"Typecheck failed, expected 'Object<{
	todos: Array<Object<{
	done: boolean,
	title: string
}>>
}>', got: (object) '{\\"todos\\":[{\\"done\\":false,\\"title\\":\\"get coffee\\"},3]}'"
`;

exports[`isShape simple 1`] = `
"Typecheck failed, expected 'Object<{
	done: boolean,
	title: string
}>', got: (object) '{}'"
`;

exports[`isShape simple 2`] = `
"Typecheck failed, expected 'Object<{
	done: boolean,
	title: string
}>', got: (object) '{\\"done\\":true,\\"title\\":3}'"
`;
